controller/src/controller.rs

```diff
@@ -105,6 +112,43 @@ impl<T: BottlerocketNodeClient> BrupopController<T> {
         None
     }
 
+    #[instrument(skip(self))]
+    fn init_metrics(&mut self) {
+        let meter = global::meter("brupop-controller");
+        self.metrics = Some(BrupopControllerMetrics::new(meter));
+    }
+
+    #[instrument(skip(self))]
+    fn emit_metrics(&self) {
+        if let Some(metrics) = &self.metrics {
+            let data = self.fetch_data();
+            metrics.emit_metrics(data);
+        }
+    }
+
+    /// Fetch the custom resources status for all resources
+    /// to gather the information on hosts's bottlerocket version
+    /// and brupop state.
+    #[instrument(skip(self))]
+    fn fetch_data(&self) -> BrupopHostsData {
+        let mut hosts_version_count_map = HashMap::new();
+        let mut hosts_state_count_map = HashMap::new();
+
+        for brn in self.all_nodes() {
+            if let Some(brn_status) = brn.status {
+                let current_version = brn_status.current_version().to_string();
+                let current_state = brn_status.current_state;
+
+                *hosts_version_count_map.entry(current_version).or_default() += 1;
+                *hosts_state_count_map
+                    .entry(serde_plain::to_string(&current_state).unwrap())
```

Comment:
Nit: We know this can never actually return an error unless we have made a programming mistake, since the enum should always be serializable. I've had some conversations with @webern about this and I think it might be best to still transform this into a `Result` type, where the `Err` variant has something like `Assertion` or some other name that implies it can only occur when the software is wrong, rather than when there's some transient error.

Happy to chat more about this!

Reply:
Fixed